File: sdks/typescript/src/universal.ts
import { redact } from "./redaction";

export const universalEventTypes = [
  "run_start",
  "run_end",
  "step_start",
  "step_end",
  "node_start",
  "node_end",
  "task_start",
  "task_end",
  "tool_start",
  "tool_end",
  "stream_delta",
  "message_delta",
  "state_checkpoint",
  "error",
  "custom",
] as const;

export type UniversalEventType = (typeof universalEventTypes)[number];

const allowedFrameworks = new Set([
  "autogen",
  "crewai",
  "custom",
  "generic",
  "langgraph",
  "microsoft-agent-framework",
  "vercel-ai",
]);
const channelRe = /^[a-z][a-z0-9_]*:[A-Za-z0-9_.:\-*]+$/;
const idRe = /^[A-Za-z0-9_.:\-]+$/;

export interface UniversalAgentEvent {
  schemaVersion: string;
  sdkName: string;
  sdkVersion: string;
  adapterName: string;
  adapterVersion: string;
  eventId: string;
  runId: string;
  threadId: string;
  agentId: string;
  framework: string;
  frameworkVersion?: string;
  language: string;
  eventType: UniversalEventType;
  channel: string;
  stepName?: string;
  nodeName?: string;
  taskName?: string;
  toolName?: string;
  input?: unknown;
  output?: unknown;
  delta?: unknown;
  messages?: unknown[];
  error?: Record<string, unknown>;
  metadata: Record<string, unknown>;
  tags: string[];
  parentCommitIds: string[];
  timestampMs: number;
}

export type UniversalAgentEventInput = Partial<UniversalAgentEvent> & Record<string, unknown>;

export function defaultThreadId(runId: string, framework: string): string {
  return `${runId}:main:${framework}`;
}

export function channelForEvent(
  event: Pick<UniversalAgentEvent, "channel" | "eventType" | "runId" | "threadId" | "framework" | "toolName">,
): string {
  if (event.channel) {
    return event.channel;
  }
  if (event.toolName) {
    return `tool:${event.runId}:${event.toolName}`;
  }
  if (event.eventType === "run_start" || event.eventType === "run_end") {
    return `framework:${event.framework}:${event.runId}`;
  }
  if (["step_end", "node_end", "task_end", "state_checkpoint"].includes(event.eventType)) {
    return `state:${event.runId}:${event.threadId}`;
  }
  return `events:${event.runId}:${event.threadId}`;
}

export function normalizeEvent(input: UniversalAgentEventInput): UniversalAgentEvent {
  const event: UniversalAgentEvent = {
    schemaVersion: stringValue(input.schemaVersion ?? input.schema_version, "schema_version", "nexus.universal.v1"),
    sdkName: stringValue(input.sdkName ?? input.sdk_name, "sdk_name", "@nexus-ipc/sdk"),
    sdkVersion: stringValue(input.sdkVersion ?? input.sdk_version, "sdk_version", "0.1.0"),
    adapterName: stringValue(input.adapterName ?? input.adapter_name, "adapter_name", "generic"),
    adapterVersion: stringValue(input.adapterVersion ?? input.adapter_version, "adapter_version", "0.1.0"),
    eventId: stringValue(input.eventId ?? input.event_id, "event_id", `evt_${randomId()}`),
    runId: stringValue(input.runId ?? input.run_id, "run_id"),
    threadId: stringValue(input.threadId ?? input.thread_id, "thread_id"),
    agentId: stringValue(input.agentId ?? input.agent_id, "agent_id"),
    framework: stringValue(input.framework, "framework"),
    frameworkVersion: optionalString(input.frameworkVersion ?? input.framework_version),
    language: stringValue(input.language, "language", "typescript"),
    eventType: eventTypeValue(input.eventType ?? input.event_type),
    channel: optionalString(input.channel) ?? "",
    stepName: optionalString(input.stepName ?? input.step_name),
    nodeName: optionalString(input.nodeName ?? input.node_name),
    taskName: optionalString(input.taskName ?? input.task_name),
    toolName: optionalString(input.toolName ?? input.tool_name),
    input: input.input,
    output: input.output,
    delta: input.delta,
    messages: optionalMessages(input.messages),
    error: optionalRecord(input.error, "error"),
    metadata: recordValue(input.metadata, "metadata", {}),
    tags: stringArray(input.tags, "tags", []),
    parentCommitIds: stringArray(input.parentCommitIds ?? input.parent_commit_ids, "parent_commit_ids", []),
    timestampMs: integerValue(input.timestampMs ?? input.timestamp_ms, "timestamp_ms", Date.now()),
  };
  event.channel = channelForEvent(event);
  validateEvent(event);
  return event;
}

export function safeJsonable(value: unknown, maxStringChars = 4000): unknown {
  if (typeof value === "string") {
    return value.length <= maxStringChars ? value : `${value.slice(0, maxStringChars)}...[truncated]`;
  }
  if (value === null || value === undefined || typeof value === "number" || typeof value === "boolean") {
    return value;
  }
  if (Array.isArray(value)) {
    return value.map((item) => safeJsonable(item, maxStringChars));
  }
  if (typeof value === "object") {
    const output: Record<string, unknown> = {};
    for (const [key, child] of Object.entries(value as Record<string, unknown>)) {
      output[key] = safeJsonable(child, maxStringChars);
    }
    return output;
  }
  if (typeof value === "function") {
    return "[function]";
  }
  return String(value);
}

export function summarizePayload(value: unknown, maxChars = 500): string {
  const text = JSON.stringify(safeJsonable(value));
  return text.length <= maxChars ? text : `${text.slice(0, maxChars)}...[truncated]`;
}

export function redactEvent(event: UniversalAgentEvent): UniversalAgentEvent {
  return redact(event).value as UniversalAgentEvent;
}

export function eventToCheckpointPayload(event: UniversalAgentEvent): Record<string, unknown> {
  const clean = redactEvent(event);
  const payload: Record<string, unknown> = {
    schema_version: clean.schemaVersion,
    sdk_name: clean.sdkName,
    sdk_version: clean.sdkVersion,
    adapter_name: clean.adapterName,
    adapter_version: clean.adapterVersion,
    event_id: clean.eventId,
    run_id: clean.runId,
    thread_id: clean.threadId,
    agent_id: clean.agentId,
    framework: clean.framework,
    language: clean.language,
    event_type: clean.eventType,
    channel: clean.channel,
    metadata: clean.metadata,
    tags: clean.tags,
    parent_commit_ids: clean.parentCommitIds,
    timestamp_ms: clean.timestampMs,
  };
  if (clean.frameworkVersion !== undefined) payload.framework_version = clean.frameworkVersion;
  if (clean.stepName !== undefined) payload.step_name = clean.stepName;
  if (clean.nodeName !== undefined) payload.node_name = clean.nodeName;
  if (clean.taskName !== undefined) payload.task_name = clean.taskName;
  if (clean.toolName !== undefined) payload.tool_name = clean.toolName;
  if (clean.input !== undefined) payload.input = clean.input;
  if (clean.output !== undefined) payload.output = clean.output;
  if (clean.delta !== undefined) payload.delta = clean.delta;
  if (clean.messages !== undefined) payload.messages = clean.messages;
  if (clean.error !== undefined) payload.error = clean.error;
  return payload;
}

function validateEvent(event: UniversalAgentEvent): void {
  if (event.schemaVersion !== "nexus.universal.v1") {
    throw new Error("UniversalAgentEvent.schema_version must be nexus.universal.v1");
  }
  for (const [field, value] of [
    ["event_id", event.eventId],
    ["run_id", event.runId],
    ["thread_id", event.threadId],
    ["agent_id", event.agentId],
  ]) {
    if (!value) {
      throw new Error(`UniversalAgentEvent.${field} is required`);
    }
    if (!idRe.test(value)) {
      throw new Error(`UniversalAgentEvent.${field} contains invalid characters`);
    }
  }
  if (!allowedFrameworks.has(event.framework)) {
    throw new Error(
      `UniversalAgentEvent.framework must be one of ${Array.from(allowedFrameworks).sort().join(", ")}`,
    );
  }
  if (!channelRe.test(event.channel)) {
    throw new Error("channel must match allowed Nexus channel pattern");
  }
  if (!Number.isInteger(event.timestampMs) || event.timestampMs <= 0) {
    throw new Error("timestamp_ms must be an integer Unix timestamp in milliseconds");
  }
}

function stringValue(value: unknown, field: string, fallback?: string): string {
  const resolved = value ?? fallback;
  if (typeof resolved !== "string" || resolved.length === 0) {
    throw new Error(`UniversalAgentEvent.${field} is required`);
  }
  return resolved;
}

function optionalString(value: unknown): string | undefined {
  if (value === undefined || value === null || value === "") {
    return undefined;
  }
  if (typeof value !== "string") {
    throw new Error("optional string fields must be strings when provided");
  }
  return value;
}

function eventTypeValue(value: unknown): UniversalEventType {
  if (typeof value !== "string" || !(universalEventTypes as readonly string[]).includes(value)) {
    throw new Error(`event_type must be one of ${universalEventTypes.join(", ")}`);
  }
  return value as UniversalEventType;
}

function recordValue(value: unknown, field: string, fallback: Record<string, unknown>): Record<string, unknown> {
  if (value === undefined || value === null) {
    return fallback;
  }
  if (typeof value !== "object" || Array.isArray(value)) {
    throw new Error(`${field} must be an object`);
  }
  return value as Record<string, unknown>;
}

function optionalRecord(value: unknown, field: string): Record<string, unknown> | undefined {
  if (value === undefined || value === null) {
    return undefined;
  }
  return recordValue(value, field, {});
}

function optionalMessages(value: unknown): unknown[] | undefined {
  if (value === undefined || value === null) {
    return undefined;
  }
  if (!Array.isArray(value)) {
    throw new Error("messages must be an array when provided");
  }
  return value;
}

function stringArray(value: unknown, field: string, fallback: string[]): string[] {
  if (value === undefined || value === null) {
    return fallback;
  }
  if (!Array.isArray(value) || !value.every((item) => typeof item === "string" && item.length > 0)) {
    throw new Error(`${field} must be an array of strings`);
  }
  return value;
}

function integerValue(value: unknown, field: string, fallback: number): number {
  const resolved = value ?? fallback;
  if (!Number.isInteger(resolved)) {
    throw new Error(`${field} must be an integer Unix timestamp in milliseconds`);
  }
  return resolved as number;
}

function randomId(): string {
  return Math.random().toString(16).slice(2) + Date.now().toString(16);
}

File: sdks/python/nexus_ipc/universal.py
from __future__ import annotations

import inspect
import json
import re
import time
import uuid
from typing import Any, Literal

from pydantic import BaseModel, Field
from pydantic import field_validator

from .redaction import redact

CHANNEL_RE = re.compile(r"^[a-z][a-z0-9_]*:[A-Za-z0-9_.:\-*]+$")
ID_RE = re.compile(r"^[A-Za-z0-9_.:\-]+$")
ALLOWED_FRAMEWORKS = {
    "autogen",
    "crewai",
    "custom",
    "generic",
    "langgraph",
    "microsoft-agent-framework",
    "vercel-ai",
}

UniversalEventType = Literal[
    "run_start",
    "run_end",
    "step_start",
    "step_end",
    "node_start",
    "node_end",
    "task_start",
    "task_end",
    "tool_start",
    "tool_end",
    "stream_delta",
    "message_delta",
    "state_checkpoint",
    "error",
    "custom",
]


class UniversalAgentEvent(BaseModel):
    schema_version: str = "nexus.universal.v1"
    sdk_name: str = "nexus-ipc"
    sdk_version: str = "0.1.0"
    adapter_name: str = "generic"
    adapter_version: str = "0.1.0"
    event_id: str = Field(default_factory=lambda: f"evt_{uuid.uuid4().hex}")
    run_id: str
    thread_id: str
    agent_id: str
    framework: str
    framework_version: str | None = None
    language: str = "python"
    event_type: UniversalEventType
    channel: str = ""
    step_name: str | None = None
    node_name: str | None = None
    task_name: str | None = None
    tool_name: str | None = None
    input: Any | None = None
    output: Any | None = None
    delta: Any | None = None
    messages: list[Any] | None = None
    error: dict[str, Any] | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    tags: list[str] = Field(default_factory=list)
    parent_commit_ids: list[str] = Field(default_factory=list)
    timestamp_ms: int = Field(default_factory=lambda: int(time.time() * 1000))

    @field_validator("schema_version")
    @classmethod
    def valid_schema_version(cls, value: str) -> str:
        if value != "nexus.universal.v1":
            raise ValueError("UniversalAgentEvent.schema_version must be nexus.universal.v1")
        return value

    @field_validator("run_id", "thread_id", "agent_id", "event_id")
    @classmethod
    def required_id(cls, value: str, info: Any) -> str:
        if not value:
            raise ValueError(f"UniversalAgentEvent.{info.field_name} is required")
        if not ID_RE.match(value):
            raise ValueError(f"UniversalAgentEvent.{info.field_name} contains invalid characters")
        return value

    @field_validator("framework")
    @classmethod
    def valid_framework(cls, value: str) -> str:
        if value not in ALLOWED_FRAMEWORKS:
            raise ValueError(
                "UniversalAgentEvent.framework must be one of "
                + ", ".join(sorted(ALLOWED_FRAMEWORKS))
            )
        return value

    @field_validator("channel")
    @classmethod
    def valid_channel(cls, value: str) -> str:
        if value and not CHANNEL_RE.match(value):
            raise ValueError("channel must match allowed Nexus channel pattern")
        return value

    @field_validator("parent_commit_ids", mode="before")
    @classmethod
    def valid_parent_commit_ids(cls, value: Any) -> list[str]:
        if value is None:
            return []
        if not isinstance(value, list) or not all(isinstance(item, str) and item for item in value):
            raise ValueError("parent_commit_ids must be an array of commit ID strings")
        return value

    @field_validator("timestamp_ms", mode="before")
    @classmethod
    def valid_timestamp_ms(cls, value: Any) -> int:
        if not isinstance(value, int):
            raise ValueError("timestamp_ms must be an integer Unix timestamp in milliseconds")
        if value <= 0:
            raise ValueError("timestamp_ms must be an integer Unix timestamp in milliseconds")
        return value


def default_thread_id(run_id: str, framework: str) -> str:
    return f"{run_id}:main:{framework}"


def channel_for_event(event: UniversalAgentEvent | dict[str, Any]) -> str:
    data = event.model_dump() if isinstance(event, UniversalAgentEvent) else event
    if data.get("channel"):
        return str(data["channel"])
    event_type = str(data.get("event_type", "custom"))
    run_id = str(data["run_id"])
    thread_id = str(data["thread_id"])
    framework = str(data["framework"])
    tool_name = data.get("tool_name")
    if tool_name:
        return f"tool:{run_id}:{tool_name}"
    if event_type in {"run_start", "run_end"}:
        return f"framework:{framework}:{run_id}"
    if event_type in {"step_end", "node_end", "task_end", "state_checkpoint"}:
        return f"state:{run_id}:{thread_id}"
    return f"events:{run_id}:{thread_id}"


def normalize_event(**kwargs: Any) -> UniversalAgentEvent:
    event = UniversalAgentEvent(**kwargs)
    if not event.channel:
        event.channel = channel_for_event(event)
    return event


def safe_jsonable(value: Any, max_string_chars: int = 4000) -> Any:
    if isinstance(value, str):
        return value if len(value) <= max_string_chars else f"{value[:max_string_chars]}...[truncated]"
    try:
        json.dumps(value)
        return value
    except TypeError:
        pass
    if hasattr(value, "model_dump") and callable(value.model_dump):
        return safe_jsonable(value.model_dump())
    if hasattr(value, "dict") and callable(value.dict):
        return safe_jsonable(value.dict())
    if isinstance(value, dict):
        return {str(key): safe_jsonable(child) for key, child in value.items()}
    if isinstance(value, list | tuple | set):
        return [safe_jsonable(child) for child in value]
    if inspect.isgenerator(value):
        return "[generator]"
    return repr(value)


def summarize_payload(value: Any, max_chars: int = 500) -> str:
    jsonable = safe_jsonable(value)
    try:
        text = json.dumps(jsonable, sort_keys=True)
    except TypeError:
        text = repr(jsonable)
    return text if len(text) <= max_chars else f"{text[:max_chars]}...[truncated]"


def redact_event(event: UniversalAgentEvent) -> UniversalAgentEvent:
    data = event.model_dump()
    clean, _ = redact(data)
    return UniversalAgentEvent.model_validate(clean)


def event_to_checkpoint_payload(event: UniversalAgentEvent) -> dict[str, Any]:
    clean = redact_event(event)
    return safe_jsonable(clean.model_dump(exclude_none=True))
